dali/operators/image/convolution/laplacian_params.h

```diff
@@ -0,0 +1,302 @@
+        prevval = val;
+      }
+      *view(cur_size - 1) = prevval;
+    }
+  }
+
+  /**
+   * @brief Derivative window of size 3 is [1, -2, 1] (which is [1, -1] composed with itself).
+   * Bigger windows are convolutions of smoothing windows with [1, -2, 1].
+   */
+  inline void PrepareDerivWindow(int window_size) {
+    for (; deriv_computed_ < window_size; deriv_computed_++) {
+      auto cur_size = deriv_computed_ + 1;
+      auto cur_idx = cur_size / 2;
+      auto &prev_view = cur_size % 2 == 0 ? smoothing_views_[cur_idx - 1] : deriv_views_[cur_idx];
+      auto &view = deriv_views_[cur_idx];
+      auto prevval = *prev_view(0);
+      *view(0) = -prevval;
+      for (int j = 1; j < cur_size - 1; j++) {
+        auto val = *prev_view(j);
+        *view(j) = prevval - *prev_view(j);
+        prevval = val;
+      }
+      *view(cur_size - 1) = prevval;
+    }
+  }
+
+  static constexpr int num_windows_ = (maxWindowSize + 1) / 2;
+  static constexpr int windows_size_ = (1 + maxWindowSize) / 2 * num_windows_;
+  int smooth_computed_, deriv_computed_;
+  std::array<T, windows_size_> smoothing_memory_;
+  std::array<T, windows_size_> deriv_memory_;
+  std::array<TensorView<StorageCPU, T, 1>, num_windows_> smoothing_views_;
+  std::array<TensorView<StorageCPU, T, 1>, num_windows_> deriv_views_;
+};
+
+
+template <int axes>
+std::array<float, axes> GetNormalizationFactors(
+    const std::array<std::array<int, axes>, axes> &window_sizes) {
+  std::array<float, axes> factors;
+  for (int i = 0; i < axes; i++) {
+    auto initial = -axes - 2;
+    auto exponent = std::accumulate(window_sizes[i].begin(), window_sizes[i].end(), initial);
+    factors[i] = exp2(-exponent);
+  }
+  return factors;
+}
+
+
+template <int axes>
+class LaplacianArgs {
+ public:
+  explicit LaplacianArgs(const OpSpec &spec)
+      : spec_{spec},
+        has_win_size_t_input_{spec.HasTensorArgument(kWindowSizeArgName)},
+        has_scale_t_input_{spec.HasTensorArgument(scaleArgName)},
+        has_scale_const_input_{spec.HasArgument(scaleArgName)},
+        normalize_{spec.GetArgument<bool>(normalizeArgName)} {}
+
+  void ObtainLaplacianArgs(const ArgumentWorkspace &ws, int nsamples) {
+    ObtainWindowSizes(ws, nsamples);
+    ObtainScales(ws, nsamples);
+  }
+
+  inline const std::array<std::array<int, axes>, axes> &GetWindowSizes(int sample_idx) {
+    return window_sizes_[sample_idx];
+  }
+
+  inline const std::array<float, axes> &GetScales(int sample_idx) {
+    return scales_[sample_idx];
+  }
+
+  inline int GetTotalWindowSizes(int sample_idx) {
+    auto &window_sizes = GetWindowSizes(sample_idx);
+    auto acc_sizes = [](int acc, const std::array<int, axes> &win_sizes) {
+      return std::accumulate(win_sizes.begin(), win_sizes.end(), acc);
+    };
+    return std::accumulate(window_sizes.begin(), window_sizes.end(), 0, acc_sizes);
+  }
+
+ private:
+  void ObtainWindowSizes(const ArgumentWorkspace &ws, int nsamples) {
```

Comment:
I revisited the ArgValue idea. API to like `HasArgumentInput` is really nice, but I am not sure if it helps a lot with different dimensionality of TensorInput data. In the end I'd like to broadcast per sample scalars, so I need to copy it sample per sample anyway. And in case of constant input I can try to avoid it batch-to-batch if batch size does not change, while calling Acquire enforces creation of TensorList of batch size every time.